Framework/API/inc/MantidAPI/SpectrumInfoIterator.h

```diff
@@ -69,18 +70,18 @@ class MANTID_API_DLL SpectrumInfoIterator
 
   void setIndex(const size_t index) { m_item.m_index = index; }
 
-  uint64_t distance_to(const SpectrumInfoIterator &other) const {
+  uint64_t distance_to(const SpectrumInfoIterator<T> &other) const {
     return static_cast<uint64_t>(other.getIndex()) -
            static_cast<uint64_t>(getIndex());
   }
 
-  bool equal(const SpectrumInfoIterator &other) const {
+  bool equal(const SpectrumInfoIterator<T> &other) const {
     return getIndex() == other.getIndex();
   }
 
-  const SpectrumInfoItem &dereference() const { return m_item; }
+  SpectrumInfoItem<T> &dereference() const { return m_item; }
 
-  SpectrumInfoItem m_item;
+  mutable SpectrumInfoItem<T> m_item;
```

Comment:
I do not like the use of `mutable` here, but I am also a bit confused: In `SpectrumInfoItem` you have a `T *m_spectrumInfo`, and a non-const method `setMasked`. Either this should not compile, or it lets you call `setMasked()` even in the `const` version. Am I missing something here?

Reply:
You instantiate what you use, you would get a compile time error if you tried to call `setMasked` on the non-const version. That might not be intuitive in the way that it would be expressed as a compiler message to the client. One option would be to use something like: ```static_assert(!std::is_const<T>::value, "Operation disabled on const T");``` to give a bit more context to compiler errors for this.

Reply:
The reason that the mutable is required is to get things working with the `iterator_facade`, that specifies a `dereference` function required on the derived class as:

```
SpectrumInfoItem<T> &dereference() const;
```
If you have a better way around this would be interested to get your ideas.

Reply:
I suppose it is ok for now.